src/BlockBuilderPolicy.sol

```diff
@@ -174,35 +188,71 @@ contract BlockBuilderPolicy is Initializable, UUPSUpgradeable, OwnableUpgradeabl
     function isAllowedPolicy(address teeAddress) public view returns (bool allowed, WorkloadId) {
+        (bool isValid, FlashtestationRegistry.RegisteredTEE memory registration) =
+            FlashtestationRegistry(registry).getRegistration(teeAddress);
+        if (!isValid) {
+            return (false, WorkloadId.wrap(0));
+        }
+
+        // @dev: could also be a separate policy registration call for gas reasons. Otherwise the builder should set gasLimit accordingly.
+        // @dev: the gas saving might not be worth the effort
+        WorkloadCacheData memory workloadData = workloadCache[registration.registrationHash];
+        if (workloadData.workloadId == 0) {
+            workloadData.tdConfigurationValid = ValidateTDConfig(registration.parsedReportBody);
+            workloadData.workloadId = WorkloadIdForTDRegistration(registration);
+            workloadCache[registration.registrationHash] = workloadData;
+        }
+
+        require(workloadData.tdConfigurationValid, "invalid td configuration");
+
         for (uint256 i = 0; i < workloadIds.length(); ++i) {
-            WorkloadId workloadId = WorkloadId.wrap(workloadIds.at(i));
-            if (FlashtestationRegistry(registry).isValidWorkload(workloadId, teeAddress)) {
-                return (true, workloadId);
+            if (workloadData.workloadId == WorkloadId.wrap(workloadIds.at(i))) {
+                return (true, workloadData.workloadId);
             }
         }
+
         return (false, WorkloadId.wrap(0));
     }
 
-    /// @notice An alternative implementation of isAllowedPolicy that verifies more than just
-    /// the workloadId's matching and if the attestation is still valid
-    /// @param teeAddress The TEE-controlled address
-    /// @param expectedTeeTcbSvn The expected teeTcbSvn of the TEE's attestation
-    /// @return allowed True if the TEE's attestation is part of the policy, is still valid, and
-    /// the teeTcbSvn matches the expected value
-    /// @dev This exists to show how different Policies can be implemented, based on what
-    /// properties of the TEE's attestation are important to verify.
-    function isAllowedPolicy2(address teeAddress, bytes16 expectedTeeTcbSvn) external view returns (bool allowed) {
-        for (uint256 i = 0; i < workloadIds.length(); ++i) {
-            WorkloadId workloadId = WorkloadId.wrap(workloadIds.at(i));
-            TD10ReportBody memory reportBody = FlashtestationRegistry(registry).getReportBody(teeAddress);
-            if (
-                FlashtestationRegistry(registry).isValidWorkload(workloadId, teeAddress)
-                    && reportBody.teeTcbSvn == expectedTeeTcbSvn
-            ) {
-                return true;
-            }
-        }
-        return false;
+    // Application specific mapping of registration data, in particular the quote and attested app data, to a workload identifier
```

Comment:
readability:

> in particular the quote and attested app data, to a workload identifier

The quote and the reportData (where the attested app data comes from) are not in here, so this comment is confusing.

Should this say something more like "Application-specific mapping of the data this policy wants to use to authorize a TEE workload"?